### PR TITLE
flatbuffers: conan v2 support

### PR DESCRIPTION
**Pull request by SpaceIm**
Specify library name and version:  **lib/1.0**

This is also a good place to share with all of us **why you are submitting this PR** (specially if it is a new addition to ConanCenter): is it a dependency of other libraries you want to package? Are you the author of the library? Thanks!

---

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/flatbuffers/all/conanfile.py [hunk 1]
```diff
@@ -57,86 +48,82 @@ def configure(self):
...
+        tc.variables["FLATBUFFERS_LIBCXX_WITH_CLANG"] = False
+        # Mimic upstream CMake/Version.cmake removed in _patch_sources()
+        version = Version(self.version)
+        tc.variables["VERSION_MAJOR"] = version.major
+        tc.variables["VERSION_MINOR"] = version.minor
+        tc.variables["VERSION_PATCH"] = version.patch
+        # For msvc shared
+        tc.variables["CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS"] = True
+        # Relocatable shared libs on Macos
+        tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0042"] = "NEW"
```
Comment (prince-chrismc) [inline recipes/flatbuffers/all/conanfile.py:86]: What do you think the advice should be for deciding on using `cache variables`? The [conan docs](https://docs.conan.io/en/latest/reference/conanfile/tools/cmake/cmaketoolchain.html?highlight=cache_variable#cache-variables) are not the most intuitive

I want to say 

> anything that needs to be set before the call to `project` or `cmake_min_version`

 but that's not super obvious -- so my current train of thought is

> whenever you use set a `CMAKE_` prefixed var

I havent seen any downside yet but Looking for a second opinion 🙏 

```suggestion
        tc.cache_variables["CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS"] = True
        # Relocatable shared libs on Macos
        tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0042"] = "NEW"
```

**Review by prince-chrismc — APPROVED** [review]

**Review by uilianries — APPROVED** [review]

File: recipes/flatbuffers/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/flatbuffers/all/conanfile.py:86]: I've explained some cases in a conan issue, but can't remember which one.

Comment (prince-chrismc) [reply recipes/flatbuffers/all/conanfile.py:86]: Thank you for the link!

**Comment by sourcedelica** [comment]
Hi - I just picked up this change in my fork of CCI and ran into issues when I tried to build with Conan 1.52.
```
flatbuffers/2.0.7: Calling build()
flatbuffers/2.0.7: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/eric/.conan/data/flatbuffers/2.0.7/_/_/build/038baac88f4c7bfa972ce5adac1616bed8fe2ef4/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/eric/.conan/data/flatbuffers/2.0.7/_/_/package/038baac88f4c7bfa972ce5adac1616bed8fe2ef4" -DCMAKE_POLICY_DEFAULT_CMP0042="NEW" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Debug" "/home/eric/.conan/data/flatbuffers/2.0.7/_/_/build/038baac88f4c7bfa972ce5adac1616bed8fe2ef4/src"
CMake Error at CMakeLists.txt:23 (project):
  VERSION ".." format invalid.


-- Configuring incomplete, errors occurred!
flatbuffers/2.0.7:
flatbuffers/2.0.7: ERROR: Package '038baac88f4c7bfa972ce5adac1616bed8fe2ef4' build failed
```
The problem is that none of the CMake variables are set in `build()` anymore.  I tried building with 
```
conan create . 2.0.7@
```
(I've added the 2.0.7 and 2.0.8 sources to my fork's `conandata.yml`).

Thanks,

**Comment by SpaceIm** [comment]
It's specific to these new versions of flatbuffers (not in CCI), so it's unrelated to conan 1.52. flatbuffers 2.0.5 recipe works fine. It's a consequence of https://github.com/google/flatbuffers/commit/0471fa807ccfb183acaac2ed9dd451a2782ea598, so will start to see such error if we package flatbuffers >= 2.0.7. And the fix is to injection VERSION_MAJOR etc with cache_variables.

**Comment by sourcedelica** [comment]
That worked - thanks!
```
diff --git a/recipes/flatbuffers/all/conanfile.py b/recipes/flatbuffers/all/conanfile.py
index ad32669a4..0ea6e0846 100644
--- a/recipes/flatbuffers/all/conanfile.py
+++ b/recipes/flatbuffers/all/conanfile.py
@@ -77,9 +77,9 @@ class FlatbuffersConan(ConanFile):
         tc.variables["FLATBUFFERS_LIBCXX_WITH_CLANG"] = False
         # Mimic upstream CMake/Version.cmake removed in _patch_sources()
         version = Version(self.version)
-        tc.variables["VERSION_MAJOR"] = version.major
-        tc.variables["VERSION_MINOR"] = version.minor
-        tc.variables["VERSION_PATCH"] = version.patch
+        tc.cache_variables["VERSION_MAJOR"] = version.major.value
+        tc.cache_variables["VERSION_MINOR"] = version.minor.value
+        tc.cache_variables["VERSION_PATCH"] = version.patch.value
         # For msvc shared
         tc.variables["CMAKE_WINDOWS_EXPORT_ALL_SYMBOLS"] = True
         # Relocatable shared libs on Macos
```

**Comment by bog-dan-ro** [comment]
This patch broke the compilation for Android armv8 ...
```
$ conan install flatbuffers/2.0.5@ --update  --build=outdated --profile:host android_armv8 --profile:build default
Configuration (profile_host):
[settings]
arch=armv8
arch_build=x86_64
build_type=Release
compiler=clang
compiler.libcxx=c++_shared
compiler.version=9
os=Android
os.api_level=21
os_build=Linux
[options]
[build_requires]
*: android-ndk/r21e
[env]

Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=12
os=Linux
os_build=Linux
[options]
[build_requires]
[env]

....
flatbuffers/2.0.5: Applying build-requirement: android-ndk/r21e
flatbuffers/2.0.5: WARN: Build folder is dirty, removing it: /home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f
flatbuffers/2.0.5: Copying sources to build folder
flatbuffers/2.0.5: Building your package in /home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f
flatbuffers/2.0.5: Generator txt created conanbuildinfo.txt
flatbuffers/2.0.5: Calling generate()
flatbuffers/2.0.5: Aggregating env generators
[HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK. 'fPIC' option found and apparently well managed
[HOOK - conan-center.py] pre_build(): [FPIC MANAGEMENT (KB-H007)] OK
flatbuffers/2.0.5: Calling build()
flatbuffers/2.0.5: CMake command: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/package/a2ccebaee30b07c93935ff3643d57c8239e9ee4f" -DCMAKE_POLICY_DEFAULT_CMP0042="NEW" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/src"
-- Using Conan toolchain: /home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/generators/conan_toolchain.cmake
-- ANDROID_PLATFORM not set. Defaulting to minimum supported version
16.
-- Conan toolchain: Setting CMAKE_POSITION_INDEPENDENT_CODE=ON (options.fPIC)
-- Conan toolchain: Setting BUILD_SHARED_LIBS = OFF
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /home/bogdan/.conan/data/android-ndk/r21e/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++
-- Check for working CXX compiler: /home/bogdan/.conan/data/android-ndk/r21e/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ - broken
CMake Error at /usr/share/cmake-3.24/Modules/CMakeTestCXXCompiler.cmake:62 (message):
  The C++ compiler

    "/home/bogdan/.conan/data/android-ndk/r21e/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/Release/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/gmake -f Makefile cmTC_6f6ed/fast && /usr/bin/gmake  -f CMakeFiles/cmTC_6f6ed.dir/build.make CMakeFiles/cmTC_6f6ed.dir/build
    gmake[1]: Entering directory '/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/Release/CMakeFiles/CMakeTmp'
    Building CXX object CMakeFiles/cmTC_6f6ed.dir/testCXXCompiler.cxx.o
    /home/bogdan/.conan/data/android-ndk/r21e/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++ --target=aarch64-none-linux-android21 --gcc-toolchain=/home/bogdan/.conan/data/android-ndk/r21e/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/toolchains/llvm/prebuilt/linux-x86_64 --sysroot=/home/bogdan/.conan/data/android-ndk/r21e/_/_/package/4db1be536558d833e52e862fd84d64d75c2b3656/toolchains/llvm/prebuilt/linux-x86_64/sysroot   -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -march=armv7-a -mthumb -Wformat -Werror=format-security    -fPIE -MD -MT CMakeFiles/cmTC_6f6ed.dir/testCXXCompiler.cxx.o -MF CMakeFiles/cmTC_6f6ed.dir/testCXXCompiler.cxx.o.d -o CMakeFiles/cmTC_6f6ed.dir/testCXXCompiler.cxx.o -c /home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/Release/CMakeFiles/CMakeTmp/testCXXCompiler.cxx
    clang++: error: the clang compiler does not support '-march=armv7-a'
    gmake[1]: *** [CMakeFiles/cmTC_6f6ed.dir/build.make:79: CMakeFiles/cmTC_6f6ed.dir/testCXXCompiler.cxx.o] Error 1
    gmake[1]: Leaving directory '/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/Release/CMakeFiles/CMakeTmp'
    gmake: *** [Makefile:127: cmTC_6f6ed/fast] Error 2
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:5 (project)


-- Configuring incomplete, errors occurred!
See also "/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/Release/CMakeFiles/CMakeOutput.log".
See also "/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/Release/CMakeFiles/CMakeError.log".
flatbuffers/2.0.5: 
flatbuffers/2.0.5: ERROR: Package 'a2ccebaee30b07c93935ff3643d57c8239e9ee4f' build failed
flatbuffers/2.0.5: WARN: Build folder /home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/Release
ERROR: flatbuffers/2.0.5: Error in build() method, line 112
        cmake.configure()
        ConanException: Error 1 while executing cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/build/generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/package/a2ccebaee30b07c93935ff3643d57c8239e9ee4f" -DCMAKE_POLICY_DEFAULT_CMP0042="NEW" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/home/bogdan/.conan/data/flatbuffers/2.0.5/_/_/build/a2ccebaee30b07c93935ff3643d57c8239e9ee4f/src"
```
Why on earth is adding `-march=armv7-a` to compiler flags  for an armv8 build?

**Comment by bog-dan-ro** [comment]
https://github.com/conan-io/conan-center-index/pull/12795 fixes it.

**Pull request merged**